stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+        match (c) {
+          -1 => {
+            let detail = "expected a key-value pair or the end of the object"
+            err = Some(buildUnexpectedTokenError(parserImplHelper, detail))
+            break
+          },
+          0x2C => { // ','
+            if (first) {
+              let detail = "expected a key-value pair or the end of the object, found ','"
+              err = Some(buildUnexpectedTokenError(parserImplHelper, detail))
+              break
+            } else {
+              next(parserImplHelper)
+              void
+            }
+          },
+          0x7D => { // '}'
+            next(parserImplHelper)
+            done = true
+            break
+          },
+          _ => {
+            // A new entry in current object.
+            // Just call parseString directly. In case the current character id not '"', it will return an error we can pass along.
+            match (parseString(parserImplHelper)) {
+              Ok(key) => {
+                skipWhiteSpace(parserImplHelper)
+
+                match (expectCodePointAndAdvance(0x3A, parserImplHelper)) {
+                  // ':'
+                  None => {
+                    // note that parseValue skips initial and final whitespace
+                    match (parseValue(parserImplHelper)) {
+                      Ok(value) => {
+                        entries = [(key, value), ...entries]
+                        first = false
+                        void
+                      },
+                      Err(e) => {
+                        err = Some(e)
+                        break
+                      },
+                    }
+                  },
+                  Some(e) => {
+                    err = Some(e)
+                    break
+                  },
+                }
+              },
+              Err(e) => {
+                err = Some(e)
+                break
+              },
+            }
+          },
+        }
+      }
+      // end of entry loop
+
+      match (err) {
+        Some(e) => Err(e),
+        None => {
+          if (done) {
+            Ok(JSONObject(List.reverse(entries)))
+          } else {
+            // This brand is not expected to actually execute,
+            // but in case it does, may just as well do the right thing.
+            Err(
+              buildUnexpectedTokenError(
+                parserImplHelper,
+                "unexpected end of object"
+              )
+            )
+          }
+        },
+      }
+    },
+  }
+}
+
+/**
+ * Parses JSON input from a string into a `JSON` object.
+ * 
+ * @param str: The JSON text string
+ * @returns A `Result` object with either the parsed `JSON` object or an error.
+ *
+ * Example output:
+ * ```
+ * Ok(JSONObject([("currency", JSONString("$")), ("price", JSONNumber(119))]))
+ * ```
+ * 
+ * @example print(parse("{\"currency\":\"$\",\"price\":119}"))
+ */
+export let parse: String -> Result<JSON, JSONParseError> = (str: String) => {
```

Comment:
We might want a `parseOpt` as well

Reply:
Do you have any options for parsing in mind?

Reply:
I can only think of options relative to number parsing. And maybe some control over internal buffer sizes, but I'd rather not expose the buffer related stuff to allow future internal refactoring.

The user could want to parse into 32 bit floats or use a lax, but fast decimal to float conversion algorithm. Probably only interesting for relatively niche use cases with enormous data sets in JSON, for which Grain wouldn't be very efficient anyway since it boxes float numbers.

**If** Grain gains a BigDecimal number type in the future, then an option would be necessary to support both and maintain compatibility with default parsing behavior to floats. Though I don't even know if BigDecimal makes sense as a subtype of Number. If not, then this issue would not apply anyway as the current `JSON` enum would not do the job. Alternatively to BigDecimal I could keep the current "hack" that uses rationals, but now with BigInt numerator/denominator.

Reply:
I think this would have been nice, another potentional option would have been `allowComments`, while it doesnt follow the spec it is something that is nice to have and could fit well into a parse option. But i think we came to the conclusion in the discord that we dont want to allow parsing options.